spec/paperclip/attachment_spec.rb

```diff
@@ -31,9 +31,109 @@
     file.close
   end
 
+  context "when generators and procesors are provided" do
+    class TestProcessor < Paperclip::Processor
+      def make
+        original_content = file.tap(&:rewind).read
+        file.rewind
+        result = Tempfile.new('paperclip-test')
+        result.print "#{original_content}#{options[:suffix]}"
+        result.tap(&:rewind)
+      end
+    end
+
+    class TestFullGenerator < Paperclip::Processor
+      def make
+        results = {}
+        original_content = file.tap(&:rewind).read
+        file.rewind
+        options.each do |name, style_options|
+          result = Tempfile.new('paperclip-test')
+          result.print "#{original_content}#{style_options[:suffix].upcase}"
+          results[name] = result.tap(&:rewind)
+        end
+        results
+      end
+    end
+
+    class TestPartGenerator < Paperclip::Processor
+      def make
+        original_content = file.tap(&:rewind).read
+        file.rewind
+        name, style_options = *options.first
+        result = Tempfile.new('paperclip-test')
+        result.print "#{original_content}#{style_options[:suffix].upcase}"
+        {name => result.tap(&:rewind)}
+      end
+    end
+
+    before do
+      Paperclip.register_processor :test_full_generator, TestFullGenerator
+      Paperclip.register_processor :test_part_generator, TestPartGenerator
+      Paperclip.register_processor :test_processor, TestProcessor
+    end
+
+    after { Paperclip.clear_processors! }
+
+    it "runs the processors and generator at the right times" do
+      rebuild_model(
+        styles: {original: {suffix: 'o'}, a: {suffix: 'a'}, b: {suffix: 'b'}},
+        generator: :test_full_generator,
+        processors: [:test_processor],
+      )
+      tempfile = Tempfile.open('paperclip-test') do |file|
+        file.print '.'
+        file
+      end
+      open(tempfile.path) do |file|
+        dummy = Dummy.new
+        dummy.avatar = file
+        dummy.save
+        assert_equal '.oAa', File.read(dummy.avatar.path(:a))
+        assert_equal '.oBb', File.read(dummy.avatar.path(:b))
+        assert_equal '.o', File.read(dummy.avatar.path(:original))
+      end
+    end
+
+    it "uses the original for styles not generated by the generator" do
+      rebuild_model(
+        styles: {original: {suffix: 'o'}, a: {suffix: 'a'}, b: {suffix: 'b'}},
```

Comment:
Space inside { missing.<br>Space inside } missing.<br>Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.